fetch.bs

```diff
@@ -1892,6 +1892,24 @@ source of security bugs. Please seek security review for features that deal with
  <a for="URL serializer"><i>exclude fragment</i></a> set to true.
 </ol>
 
+<p>To check <dfn export>Cross-Origin-Embedder-Policy allows credentials</dfn>, given a
+<a for=/>request</a> <var>request</var>, run theses steps:
+
+<ol>
+ <li><p>If <var>request</var>'s <a for=request>mode</a> is not <code>no-cors</code>", return
+ true.</p>
+
+ <li><p>If <var>request</var>'s <a for=request>client</a> is null, return true.</p>
+
+ <li><p>If <var>request</var>'s <a for=request>client</a>'s <a for="environment settings
+ object">embedder policy</a> is not
+ "<code><a for="embedder policy value">credentialless</a></code>", return true.</p>
+
+ <li><p>If <var>request</var>'s <a for=request>origin</a> is <a>same origin</a> with
+ <var>request</var>'s <a for=request>current URL</a>'s <a for=url>origin</a>, return true.</p>
```

Comment:
This check seems sketchy. What about redirects? Should this use tainted origin flag instead?

Reply:
It was described by: https://wicg.github.io/credentiallessness/#redirect

> The decision to include credentials is done independently for each request. The variable includeCredentials is set for the initial request, but also after each redirect.
>
> For example, credentials are not included for a cross-origin no-cors request, but they can be added in the next request if it redirects to a same-origin resource.

Tested by: [redirect.tentative.html](https://wpt.fyi/results/html/cross-origin-embedder-policy/credentialless/redirect.tentative.html?label=master&label=experimental&aligned&q=redirect.tentative.html)

The tainted origin flag, could be used to omit credentials whenever there is a single cross-origin URL in the redirect chain. However, that's not what was originally envisioned. Do you think there is an argument if favor of a different behavior?

Reply:
It means that in A -> B -> A, the final A isn't considered cross-origin, which is generally a thing we consider a bug, due to confused deputy attacks. It's considered cross-origin for CORS, tainting of responses, etc.

Reply:
It was made this way to avoid breaking more clients than necessary. Goal is to make COEP deployable easily.

In the threat model, the attacker wants to use SAB to read a non-public cross-origin resource. So, allowing credentials for same-origin resource doesn't go against it.

However, I agree doing something similar to other logic would bring some kind of purity here. That's good too.

Do you feel strongly against it? This probably deserve some attention from @mikewest or @arturjanc with whom we discussed it in the design doc.

Reply:
I don't remember discussing this in particular, so apologies if I disagree with my past self: at the moment, I think it's probably best to match our behavior for `Sec-Fetch-Site` which will treat a request as `cross-origin` for A->B->A transitions (https://w3c.github.io/webappsec-fetch-metadata/#:~:text=for%20each%20url%20in%20r%E2%80%99s%20url%20list%3A). We didn't use the tainted origin flag there, both because I didn't know about it (TIL!), and because we needed to distinguish `cross-origin` and `same-site`.

It might be reasonable to use here, since we don't care about the site boundary in this mechanism.

Reply:
If you both feel this is reasonable thing to do, then I guess... let's do it.
It was actually way simpler to implement this way.

However, I haven't seen a real security reason to do it here. We can ship with this strict version and consider relaxing it later, if we discover this creates some difficulties with developers.

Reply:
The security issue is explained here: https://en.wikipedia.org/wiki/Confused_deputy_problem. B gets to make a request against A with the authority of A if you do not consider it to be cross-origin.

Reply:
annevk@, I addressed your suggestion. In the latest patchset, we take into account the tainted origin flag.

This gives:
![Capture d’écran du 2021-09-20 15-11-03](https://user-images.githubusercontent.com/4759106/134008757-c834ce74-091e-49b4-abde-5dae072efad0.png)

Note: Thanks for the link about "Confused deputy problem", however I still don't see there are anything about `COEP:credentialless ` and the threat model about COI & spectre attacks where this really matter.

Reply:
(The corresponding WPT/chrome [change](https://github.com/web-platform-tests/wpt/pull/30871/files))

Reply:
It's not a Spectre-related threat. Say you adopt credentialless to avoid sending credentials to other origins. You might reasonably expect that only same-origin requests contain credentials. But you might not expect that those other origins can also make same-origin requests through redirects. And they really shouldn't be able to do that as they don't have your authority.